### PR TITLE
Checkout: phone number validation fix #27021

### LINKED ISSUE #27021: On checkout, entering an invalid phone number incorrectly marks the input as valid
**Prerequisites:**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
On checkout, entering an invalid phone number will mark the phone as valid (green border) instead of as invalid (red border).

**Expected behavior**
If phone number is invalid it should get the styling of invalid field (red border)

**Actual behavior**
Invalid phone number gets styling of a valid field (green border)

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Add product to cart
2. Go to checkout
3. Fill an invalid phone number, for example insert: `a123`
4. See an error message appears indicating phone number is invalid, yet the phone input is marked as valid (green border)

**Screenshots**
image
image

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
Can be seen also in [Storefront](https://themes.woocommerce.com/storefront/)

### PR DESCRIPTION
**Pull request by jgreys**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #27021

### How to test the changes in this Pull Request:

1. Add product to cart
2. Go to checkout
3. Fill an invalid phone number, for example insert: a123
4. Submit
5. The phone input is marked as invalid (red border)

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Tweak - Introduced phone number input validation.

### CODE REVIEW

File: assets/js/frontend/checkout.js [hunk 1]
```diff
@@ -252,14 +253,23 @@ jQuery( function( $ ) {
...
 						if ( ! pattern.test( $this.val()  ) ) {
-							$parent.removeClass( 'woocommerce-validated' ).addClass( 'woocommerce-invalid woocommerce-invalid-email' );
+							$parent.removeClass( 'woocommerce-validated' ).addClass( 'woocommerce-invalid woocommerce-invalid-email woocommerce-invalid-phone' );
 							validated = false;
 						}
 					}
 				}
 
+				if ( validate_phone ) {
+					var pattern = new RegExp(/[\s\#0-9_\-\+\/\(\)\.]/g);
```
Comment (rrennick) [inline assets/js/frontend/checkout.js:263]: @jgreys Thanks for the pull request. `var pattern` is a duplicate declaration of `pattern`. You'll need to move the declaration to the beginning of the function and remove the `var` on line #254.

**Review by rrennick — DISMISSED** [review]
This PR worked fine with the above issue addressed.

**Comment by rrennick** [comment]
@jgreys Thanks, I've added it to the review queue.

**Comment by mi5t4n** [comment]
hhgfhg

File: assets/js/frontend/checkout.js [hunk 2]
```diff
@@ -225,17 +227,26 @@ jQuery( function( $ ) {
...
+						pattern = new RegExp(/^([a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+(\.[a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+)*|"((([ \t]*\r\n)?[ \t]+)?([\x01-\x08\x0b\x0c\x0e-\x1f\x7f\x21\x23-\x5b\x5d-\x7e\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|\\[\x01-\x09\x0b\x0c\x0d-\x7f\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))*(([ \t]*\r\n)?[ \t]+)?")@(([a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.)+([a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[0-9a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.?$/i); // eslint-disable-line max-len
```
Comment (claudiosanches) [inline assets/js/frontend/checkout.js:230]:
```suggestion
						pattern = new RegExp( /^([a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+(\.[a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+)*|"((([ \t]*\r\n)?[ \t]+)?([\x01-\x08\x0b\x0c\x0e-\x1f\x7f\x21\x23-\x5b\x5d-\x7e\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|\\[\x01-\x09\x0b\x0c\x0d-\x7f\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))*(([ \t]*\r\n)?[ \t]+)?")@(([a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.)+([a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[0-9a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.?$/i ); // eslint-disable-line max-len
```

File: assets/js/frontend/checkout.js [hunk 3]
```diff
@@ -225,17 +227,26 @@ jQuery( function( $ ) {
...
 						if ( ! pattern.test( $this.val()  ) ) {
-							$parent.removeClass( 'woocommerce-validated' ).addClass( 'woocommerce-invalid woocommerce-invalid-email' );
+							$parent.removeClass( 'woocommerce-validated' ).addClass( 'woocommerce-invalid woocommerce-invalid-email woocommerce-invalid-phone' );
 							validated = false;
 						}
 					}
 				}
 
+				if ( validate_phone ) {
+					pattern = new RegExp(/[\s\#0-9_\-\+\/\(\)\.]/g);
```
Comment (claudiosanches) [inline assets/js/frontend/checkout.js:240]:
```suggestion
					pattern = new RegExp( /[\s\#0-9_\-\+\/\(\)\.]/g );
```

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Hello, there's some coding standards fixes that needs to get applied.

**Comment by jgreys** [comment]
Hello, do you need me to change anything else? Is it still in review? 🙂

**Review by claudiosanches — APPROVED** [review]
Thanks for your contribution!

**Comment by senadir** [comment]
Sorry for reaching out to an old PR but why was `woocommece-invalid-phone` added to email field?

**Pull request merged**